Add fourth Zephyr blog post

scottslowe opened the pull request:
This PR adds a Zephyr blog post about using stack references to connect multiple Pulumi projects.

scottslowe commented:
Closes https://github.com/pulumi/devrel-team/issues/526 when merged

jkodroff commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+As you may have read in previous Zephyr posts, the ultimate goal of the Zephyr series is to share recommended practices on the use of Pulumi to manage your infrastructure and application resources (using our fictional company and a fairly complex containerized application as the use case). However, the series exposes those recommended practices over time---not all right away, and not without also ...
+
+Here are links to all the blog posts in the series (entries below that are not linked are planned but haven't yet been published---this will get updated as new posts are published):
```
```suggestion
Here are links to all the blog posts in the series (entries below that are not linked are planned but haven't yet been published and will be updated as new posts are published):
```

scottslowe replied:
Addressed in f28ec13c

jkodroff commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+## Reviewing Zephyr's current status
+
+As this post is somewhat of a continuation of the previous post, Zephyr's use of Pulumi remains the same as it did at the end of the third post. That is, Zephyr now has _three_ Pulumi projects (zephyr-infra, zephyr-k8s, and zephyr-app). Each of these projects handles a different aspect or portion of the overall set of requirements for Zephyr's online store:
```
```suggestion
This post is a continuation of the previous post in the series: Zephyr's infrastructure footprint and Pulumi codebase remains the same, that is, Zephyr now has _three_ Pulumi projects (zephyr-infra, zephyr-k8s, and zephyr-app). Each of these projects handles a different aspect of the infrastructure for Zephyr's online store:
```

scottslowe replied:
Incorporated in f23ec13c

jkodroff commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+## Linking stacks
+
+For those that are unfamiliar, stack references are a feature in Pulumi that allows you to access output values from one stack in another stack. In other words, stack references provide a way to programmatically "link" stacks in different projects for the purpose of sharing information. You can find more details on stack references [here in the Pulumi docs](/docs/intro/concepts/stack/#stackrefere ...
```
```suggestion
Stack references are a feature in Pulumi that allow you to access output values from one stack in another stack. In other words, stack references provide a way to programmatically "link" stacks in different projects for the purpose of sharing information. You can find more details on stack references [in the Pulumi docs](/docs/intro/concepts/stack/#stackreferences).
```

scottslowe replied:
Partially addressed in f28ec13c. My English teacher wife and I agree that "allows" is correct, but I'll likely reword this in a future change just to avoid any confusion or disagreement. :-)

scottslowe replied:
Reworded completely in 17aacd6f

jkodroff commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+Using stack references, the Zephyr team could make the necessary connections they needed:
+
+* The code handling the Kubernetes cluster could reference the VPC ID and subnet IDs from the base infrastructure layer.
```
```suggestion
* The code handling the Kubernetes cluster can reference the VPC ID and subnet IDs from the base infrastructure layer.
```

scottslowe replied:
Addressed in f28ec13c

jkodroff commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+Using stack references, the Zephyr team could make the necessary connections they needed:
+
+* The code handling the Kubernetes cluster could reference the VPC ID and subnet IDs from the base infrastructure layer.
+* Similarly, the code that was responsible for deploying the online store services could reference the necessary Kubernetes connection details for the Kubernetes cluster.
```
```suggestion
* Similarly, the code that is responsible for deploying the online store services can reference the necessary Kubernetes connection details (kubeconfig) for the Kubernetes cluster.
```

scottslowe replied:
Fixed in f28ec13c

jkodroff commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+* The code handling the Kubernetes cluster could reference the VPC ID and subnet IDs from the base infrastructure layer.
+* Similarly, the code that was responsible for deploying the online store services could reference the necessary Kubernetes connection details for the Kubernetes cluster.
+
+It's worthwhile to note that any information that needs to be accessible from another stack via a stack reference must be exported as a stack output in the source stack. If you don't mark it as a stack output, then it can't be used in a stack reference. Fortunately, adding stack outputs after the fact is easy and generally has no impact whatsoever on existing infrastructure.
```
We should either omit "generally" or explain the exceptions here. (I believe it never has an effect, but maybe there's something I'm not remembering.)

scottslowe replied:
Done, taken care of in f28ec13c

jkodroff commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+It's worthwhile to note that any information that needs to be accessible from another stack via a stack reference must be exported as a stack output in the source stack. If you don't mark it as a stack output, then it can't be used in a stack reference. Fortunately, adding stack outputs after the fact is easy and generally has no impact whatsoever on existing infrastructure.
+
+Additionally, any value retrieved via a stack reference is treated as an `Output`, and therefore may require some additional work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps enormously in this situation. Some SDKs also have language-specific me ...
```
```suggestion
Additionally, any value retrieved via a stack reference is treated as a Pulumi Output, and therefore may require some additional work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps enormously in this situation. Some SDKs also have language-specific mechanisms that can help; for example, using Go's `.AsStringArrayOutput()` method on a `StackReference.GetOutput` statement makes referencing subnet IDs from other project much easier.
```

scottslowe replied:
Handled by f28ec13c

jkodroff commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+Additionally, any value retrieved via a stack reference is treated as an `Output`, and therefore may require some additional work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps enormously in this situation. Some SDKs also have language-specific me ...
+
+While stack references are generally straightforward, there are some recommended practices to observe around the use of stack outputs and stack references:
```
```suggestion
While stack references are conceptually straightforward and easy to implement, there are some recommended practices to be mindful of regarding the use of stack outputs and stack references:
```

jkodroff commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+1. With regard to stack outputs, **only export what is needed.** If there's a need for it to be accessed from outside the stack, export it; otherwise, don't. It's easy to add more stack outputs after the fact, and as has been mentioned already this is done with a quick `pulumi up` that has no affect on existing infrastructure (you're only modifying the stack object itself).
+2. **Be judicious in the use of stack references.** Each stack reference is a call to the backend to read a value, which adds some level of latency. Using lots of stack references (think more than 20 or so) can introduce notable latency in Pulumi operations.
+3. **If you need lots of stack references, use a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in only a single call to the backend. Be aware, however, that you'll need to write the necessary code to understand/impo ...
+4. In line with parameterizing as much of your code as possible, **also be sure to parameterize your stack references.** A stack reference is built using an organization name, a project name, and a stack name. Don't hardcode these values; instead, use configuration values to allow the users to specify from which source stack(s) the values will be referenced.
```
I can't think of a use case where organization would need to be templates, but stack name definitely should be. It might be nice to show a default value that is the current stack name, e.g.

```
const stack = new pulumi.Config().getStack().split("/")[2];
const stackRef = new pulumi.StackReference(`my-organization/workshop-aws-proton-base-infra/${stack}`);
```

Double check the above code in an actual program before publishing - I can't vouch for its correctness, but it's something like that.

scottslowe replied:
To me, it seems entirely plausible that a company could have two organizations, with users that are members of both organizations, and there be a need to read information across the organizational boundary via a stack reference (two different business units, perhaps?).

cnunciato replied:
+1 to add some code here to show what "parameterize your stack references" looks like.

scottslowe replied:
Code added in 17aacd6f, along with references to the code in other parts of the post

scottslowe commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+This is the fourth post in a series of blog posts focused on Zephyr Archaeotech Emporium---our fictional company---and their use of Pulumi to manage their online retail store. In the first three posts, you saw how Zephyr's initial use of Pulumi changed as the company grew, and how the use of short-lived per-developer stacks helped Zephyr's application development team meet the demands of a rapidl ...
+
+As you may have read in previous Zephyr posts, the ultimate goal of the Zephyr series is to share recommended practices on the use of Pulumi to manage your infrastructure and application resources (using our fictional company and a fairly complex containerized application as the use case). However, the series exposes those recommended practices over time---not all right away, and not without also ...
```
Which portions, in particular, create that impression? (That will help me target the rewording appropriately.)

scottslowe commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+As this post is somewhat of a continuation of the previous post, Zephyr's use of Pulumi remains the same as it did at the end of the third post. That is, Zephyr now has _three_ Pulumi projects (zephyr-infra, zephyr-k8s, and zephyr-app). Each of these projects handles a different aspect or portion of the overall set of requirements for Zephyr's online store:
+
+* The "zephyr-infra" project handles base infrastructure (VPC and related constructs).
```
It's too late to rename the repo now, but it's not to say that we can't show a justification for renaming the project (and the repo) later (this is a supported operation, after all). As for why the 3 repos/projects...well, that was the point of the previous post. :-) I can briefly summarize that again here just for the sake of completeness.

scottslowe replied:
Fixed in f28ec13c

cnunciato replied:
I do think it would be nice to link these projects to their respective repos, though. What do you think? Should we not?

scottslowe commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+While stack references are generally straightforward, there are some recommended practices to observe around the use of stack outputs and stack references:
+
+1. With regard to stack outputs, **only export what is needed.** If there's a need for it to be accessed from outside the stack, export it; otherwise, don't. It's easy to add more stack outputs after the fact, and as has been mentioned already this is done with a quick `pulumi up` that has no affect on existing infrastructure (you're only modifying the stack object itself).
+2. **Be judicious in the use of stack references.** Each stack reference is a call to the backend to read a value, which adds some level of latency. Using lots of stack references (think more than 20 or so) can introduce notable latency in Pulumi operations.
```
@ringods @pierskarsenbarg Since you folks brought up the concern regarding minimize stack references and calls to read values, can you confirm if the concern is in stack outputs or stack references? Based on the comments in Slack, I believe it is the reading stack (and therefore stack references are the concern).

scottslowe replied:
Resolved via Slack discussions (see threads in `#project-zephyr` and `#questions`).

scottslowe commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+1. With regard to stack outputs, **only export what is needed.** If there's a need for it to be accessed from outside the stack, export it; otherwise, don't. It's easy to add more stack outputs after the fact, and as has been mentioned already this is done with a quick `pulumi up` that has no affect on existing infrastructure (you're only modifying the stack object itself).
+2. **Be judicious in the use of stack references.** Each stack reference is a call to the backend to read a value, which adds some level of latency. Using lots of stack references (think more than 20 or so) can introduce notable latency in Pulumi operations.
+3. **If you need lots of stack references, use a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in only a single call to the backend. Be aware, however, that you'll need to write the necessary code to understand/impo ...
```
@ringods @pierskarsenbarg Pinging you two again here. I read the exchange in Slack as needing to build an interface in the reading stack, but I could certainly see the need to build an interface/struct in the source stack as well. Can you confirm based on your understanding?

scottslowe replied:
After some lengthy Slack discussions, I think this has been resolved.

scottslowe commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+3. **If you need lots of stack references, use a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in only a single call to the backend. Be aware, however, that you'll need to write the necessary code to understand/impo ...
+4. In line with parameterizing as much of your code as possible, **also be sure to parameterize your stack references.** A stack reference is built using an organization name, a project name, and a stack name. Don't hardcode these values; instead, use configuration values to allow the users to specify from which source stack(s) the values will be referenced.
+
+## Examining Zephyr's use of stack references
```
I'd really like readers to be able to see the highlighted code in the context of the broader Pulumi program. How strongly do you feel about putting code inline into the blog post?

scottslowe replied:
Code samples added in 17aacd6f

scottslowe commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+It's clear that there are key dependencies across projects. The "zephyr-k8s" project needs to know the VPC ID and the subnet IDs from the base infrastructure project. Similarly, the "zephyr-app" project needs to have access to the Kubernetes cluster details from the "zephyr-k8s" project. As the Zephyr team went about building the code that is now in use in their multi-project architecture, they n ...
+
+## Linking stacks
```
Addressed in f28ec13c

scottslowe commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+For those that are unfamiliar, stack references are a feature in Pulumi that allows you to access output values from one stack in another stack. In other words, stack references provide a way to programmatically "link" stacks in different projects for the purpose of sharing information. You can find more details on stack references [here in the Pulumi docs](/docs/intro/concepts/stack/#stackrefere ...
+
+Using stack references, the Zephyr team could make the necessary connections they needed:
```
Fixed in f28ec13c

scottslowe commented:
@jkodroff Thanks for the detailed feedback. I believe I've addressed most of your points; there are still a couple outstanding, but I'll continue to work on them (I've responded to your comments, so if you have additional feedback in those areas I'd love to hear).

cnunciato commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,89 @@
+
+While stack references are conceptually straightforward and not difficult to implement, there are some recommended practices to be mindful of regarding the use of stack outputs and stack references:
+
+1. When it comes to stack outputs, **export what is needed.** If it needs to be accessed from outside the stack, export it; otherwise, don't. It's simple to add more stack outputs after the fact, and as has been mentioned already this is done with a quick `pulumi up` that has no affect on existing infrastructure (you're modifying the stack object itself).
```
>If it needs to be accessed from outside the stack, export it; otherwise, don't.

It occurred to me that after reading this, I might wonder, Why not? Is there any harm in doing so? There probably isn't, unless what you're exporting is somehow potentially sensitive of course. The "why" here might therefore be something along the lines of the same "why" that encourages encapsulation in a public API: expose what the API user cares about or is likely to need, hide all the rest as implementation detail (which ultimately leaves you with more ability to change the underlying implementation without affecting the API).

scottslowe replied:
I've added some text, borrowing from your suggestion, in 17aacd6f

cnunciato commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,89 @@
+
+1. When it comes to stack outputs, **export what is needed.** If it needs to be accessed from outside the stack, export it; otherwise, don't. It's simple to add more stack outputs after the fact, and as has been mentioned already this is done with a quick `pulumi up` that has no affect on existing infrastructure (you're modifying the stack object itself).
+2. **Be judicious in the use of stack references.** Each stack reference is a call to the backend to read a value, which adds some level of latency. Using lots of stack references (think more than 20 or so) can introduce notable latency in Pulumi operations.
+3. **If you need lots of stack references, use a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in a single call to the backend. Be aware, though, that you'll need to write the necessary code to understand/import/unm ...
```
I might tweak this slightly to clarify we're talking about the exporting, not the consuming (but up to you -- feel free to ignore if you don't think it adds anything):

```suggestion
3. **If you need to expose lots of outputs on a stack, consider exporting a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in a single call to the backend. Be aware, though, that you'll need to write the necessary code to understand/import/unmarshall that JSON object in the referring stack, so there is a small amount of extra work required in this situation.
```

It might also be worth mentioning that this approach can also make it a little easier to codify the API contract between stacks and their consumers (vs. that contract consisting of "a bunch of optional strings"). 

Relatedly, we might want to mention that outputs are expected to be primitive/serializable types -- string, bool, object, etc. Occasionally folks think they can export whole resource instances (as you would in a typical program) and then realize this doesn't quite work, and are confused. (The docs may also mention this, though.)

scottslowe replied:
In 17aacd6f I've reworded per your suggestion, but there is still an outstanding question as to whether it is the stack outputs that are "costly" or if it is the stack references reading those outputs that's "costly" (from a latency perspective). I've asked @ringods & @pierskarsenbarg to weigh in on that, as I'm not sure.

scottslowe replied:
I think this has been resolved (refer to Slack discussions in #project-zephyr and #questions).

cnunciato reviewed (commented):
Looking good! Some suggestions for you to consider.

susanev commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,87 @@
+
+It's worthwhile to note that any information that needs to be accessible from another stack via a stack reference must be exported as a stack output in the source stack. If you don't mark it as a stack output, then it can't be used in a stack reference. Fortunately, adding stack outputs after the fact is easy and generally has no impact whatsoever on existing infrastructure.
+
+Additionally, any value retrieved via a stack reference is treated as an `Output`, and therefore may require some additional work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps enormously in this situation. Some SDKs also have language-specific me ...
```
we also have this in our style guide https://github.com/pulumi/pulumi-hugo/blob/master/STYLE-GUIDE.md

scottslowe replied:
Fixed in 17aacd6f

scottslowe commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,89 @@
+* Finally, in the application stack, the [stack reference values are again parameterized](https://github.com/pulumi/zephyr-app/blob/multi-project/infra/index.ts#L7-L9), and the Kubeconfig---necessary for the application stack to deploy onto the provisioned Kubernetes cluster---is [referenced via a stack reference](https://github.com/pulumi/zephyr-app/blob/multi-project/infra/index.ts#L11-L13).
+
+{{% notes %}}
+All of the GitHub links in the paragraphs above reference the `multi-project` tag in each repository, which has the code as of this blog post (and the earlier blog post).
```
> I link to link these references
To which references are you referring? The blog post references? Or the GitHub references? (I think the latter, if I'm reading your comment correctly.)

cnunciato replied:
You know what, I'm sorry -- I must've misread this. I think I read it as "The code above refers to the `multi-project` tag in the repository..." so my suggestion was just to link "multi-project tag" to the tag in the repo. :ignoreme:

scottslowe commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,89 @@
+
+Using stack references, the Zephyr team can make the necessary connections they needed:
+
+* The code handling the Kubernetes cluster can reference the VPC ID and subnet IDs from the base infrastructure layer.
+* Similarly, the code that is responsible for deploying the online store services can reference the necessary Kubernetes connection details (the Kubeconfig) for the Kubernetes cluster.
```
Code was added toward the end in 17aacd6f and I've put a reference to that code here

scottslowe commented on themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:
```diff
@@ -0,0 +1,89 @@
+
+It's worthwhile to note that any information that needs to be accessible from another stack via a stack reference must be exported as a stack output in the source stack. If you don't mark it as a stack output, then it can't be used in a stack reference. Adding stack outputs after the fact requires little effort and has no impact on existing infrastructure.
+
+Also, any value retrieved via a stack reference is treated as a Pulumi Output, and therefore may require some extra work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps considerably in this situation. Some SDKs also have language-specific mechanism ...
```
Done, fixed in 17aacd6f

cnunciato reviewed (approved):
Nice! Somehow it feels like it ends a bit quickly (I was all into following those code snippets!), but that's okay -- if there's nothing more to be said, there's nothing more to be said! 😄 The snippets totally help, though -- thanks for adding those. 

Other than a couple of minor suggestions to add some links and pull in that sweet graphic you made showing the stacks and their relationships (and I guess adding a meta image as well before merging), this looks good to me. 👍

ringods commented on themes/default/content/blog/iac-recommended-practices-using-stack-references/index.md:
```diff
@@ -0,0 +1,132 @@
+3. **If you need to expose lots of outputs on a stack, consider exporting a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in a single call to the backend. Be aware, though, that you'll need to write the necessary co ...
+4. In line with parameterizing as much of your code as possible, **also be sure to parameterize your stack references.** A stack reference is built using an organization name, a project name, and a stack name. Don't hardcode these values; instead, use configuration values to allow the users to specify from which source stack(s) the values will be referenced. You'll see an example of that in the n ...
+5. **Plan for handling cascading stack updates.** If a stack has outputs being consumed via stack references, what happens when those stack outputs change? The Pulumi Service [supports webhooks](/docs/intro/pulumi-service/webhooks/) that can help, and Pulumi is currently exploring options for having Pulumi Deployments help with this (see [this GitHub issue](https://github.com/pulumi/service-reque ...
+
```
I would add a 6th bullet point talking about security. With separate stacks also comes that not everyone should be allowed to access the stack outputs of the production stack. I would highlight [Stack Permissions](https://www.pulumi.com/docs/intro/pulumi-service/projects-and-stacks/#stack-permissions) and [Teams and RBAC](https://www.pulumi.com/docs/intro/pulumi-service/teams/) to control such access.

scottslowe replied:
This is an excellent suggestion as a revision to the previous post, which discusses _why_ you would use multiple projects. Unless I'm misreading your suggestion, it sounds like it would fall squarely in that post as opposed to this one (which is focused on the use of stack references).

scottslowe replied:
I've created https://github.com/pulumi/devrel-team/issues/530 to track this work.

ringods replied:
I would integrate it here. After splitting the teams (article 3), you are now introducing the stack references. If you layer projects (infra, k8s, app), you can also talk about stack permissions explaining that the access to the infra and/or k8s is read-only to the app team, while read-write for the infra team.

If you export secrets in a stack, you definitely want to limit access to these, hence also my suggestion (in this article) to narrow the permissions overall to the production stacks of each Pulumi program.

scottslowe replied:
Keep in mind that the next post in the series (after we take a quick break to look at local testing) is security and RBAC, in which we will most definitely look at this in more detail. I'll see if I can work in a quick mention here with a statement that we'll be exploring the topic in more detail in an upcoming post.

scottslowe replied:
Fixed in b34f9cd9

ringods reviewed (changes requested):
@scottslowe in the beginning of each article, you have the overview of all articles in the series. In the first 3 articles, you didn't activate the link to this 4th article yet.

scottslowe commented:
> @scottslowe in the beginning of each article, you have the overview of all articles in the series. In the first 3 articles, you didn't activate the link to this 4th article yet.

This is addressed in b34f9cd9, thanks!

scottslowe commented:
Just waiting on final meta image before we can merge.

cnunciato reviewed (approved):
Nice work -- looks great. 🚀
